File: contains-duplicate/contains-duplicate-linearithmic-time-v2/ContainsDuplicate.cpp
// https://leetcode.com/problems/contains-duplicate
// Linearithmic time, linear extra-space.
// Concepts: Balanced trees, Counting.
class ContainsDuplicate {
public:
    bool containsDuplicate(vector<int>& nums) {
        map<int,int> counts;
        for (int num : nums) {
            counts[num] ++;
        }
        for (auto [_, count] : counts) {
            if (count > 1) {
                return true;
            }
        }
        return false;
    }
};

File: contains-duplicate/contains-duplicate-quadratic-time/ContainsDuplicate.java
// https://leetcode.com/problems/contains-duplicate
// LC217. Contains Duplicate
// Quadratic time, constant extra-space.
class ContainsDuplicate {
    public boolean containsDuplicate(int[] nums) {
        int n = nums.length;
        for (int i=0; i<n; i++) {
            for (int j=i+1; j<n; j++) {
                if (nums[i] == nums[j]) {
                    return true;
                }
            }
        }
        return false;
    }
}
